yapapi/script/__init__.py

```diff
@@ -81,9 +80,15 @@ def process_batch_event(
             if event.success:
                 command._result.set_result(event)
             else:
-                command_str = str(command.evaluate(self._ctx))  # TODO -> Command.`__repr__`
-                raise CommandExecutionError(command_str, event.message, event.stderr)
-        return event  # type: ignore # -> TODO #786
+                raise CommandExecutionError(str(command), event.message, event.stderr)
+
+    @property
+    def results(self) -> List[CommandExecuted]:
```

Comment:
Just wondering - wouldn't it be more in-line with what we already have to make this method `async` and await all the commands' results?
We can currently await for the result of any single script step (see: `Script#add`), so it would be consistent for a method which returns all results to be async as well.
It won't change how this function behaves in the call sites we already have, but it might be more intuitive in future use (having a method which needs to be explicitly awaited vs getting a somewhat obscure exception).

Reply:
1. Since yyour review I've pushed one more commit that removes the "internal" usage of this property, so we can without any additional cost modify this any way we like. Or also we can just remove it :)

2. The awaitable thingy with all results is available via
```
future_results = yield script
await future_results
```
--> e.g. check the `yacat` example. I don't see a purpose of having a `Script` function that just repeats exactly the same (maybe there is a reason, but I don't know any).

3. About the "obscure" exceptions: OK, I get your point. For me, the logic behind this implementation looks as follows:
* there are some sections of code where we know this `Script` is already finished
* using `await` on objects when it's obvious nothing is really `await`ing for anything just makes the code less readable
* --> this sort of a property, when used in a correct way, makes things easier.

I've a feeling we already had a similar discussion about `yapapi.rest.market.Agreement.cached_details` :)

Reply:
TL;DR 

I'm against `asyncing` this method because such functionality is already available.
If you think we'll do better without such possibly-confusing method, I can agree and remove it.

Reply:
@zakaprov @johny-b `results()` can also return an async generator which would yield command results one by one.

Reply:
@johny-b Fair point, I missed that we can already await on the `Script` instance as such. Fine with me, let's keep `results` as is. Same goes for @filipgolem's comment - I don't think we need yet another way of retrieving results in an async manner if we already have a future from the `Script` itself, as well as futures for its individual steps.

Reply:
OK, so I'm keeping it the way it is.

Also, about "another way of retrieving results in an async matter" and @filipgolem's idea: all results are `Event`s, so to process events in an async-generator you could do something like:

```
results_queue = asyncio.queue()

async def results_gen():
    for event in results_queue:
        yield event

def event_consumer(event):
    if isinstance(event, CommandEvent):
        results_queue.push_nowait(event)
        
Golem.add_event_consumer(event_consumer)
```

and just never have any "result consuming" in your worker!

Reply:
(I'm not saying this is a good idea, but in general I think we're rather on the "too many" than "too little" side when talking about result retrieving).